src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/waypoints_pending_resumption_handler.cc

```diff
@@ -36,99 +36,104 @@ namespace sdl_rpc_plugin {
+  const auto corr_id = get_corr_id_from_message(*request);
   auto resumption_request =
-      MakeResumptionRequest(corr_id, function_id, *request);
-  app_ids_.push(app.app_id());
+      MakeResumptionRequest(corr_id, Navigation_SubscribeWayPoints, *request);
 
-  if (pending_requests_.empty()) {
-    SDL_LOG_DEBUG("There are no pending requests for app_id: " << app.app_id());
-    pending_requests_[corr_id] = request_ref;
-    subscribe_on_event(function_id, corr_id);
-    SDL_LOG_DEBUG("Sending request with function id: "
-                  << function_id << " and correlation_id: " << corr_id);
-
-    application_manager_.GetRPCService().ManageHMICommand(request);
-  } else {
-    SDL_LOG_DEBUG("There are pending requests. Frozen resumption for app id "
-                  << app.app_id() << " corr id = " << corr_id);
-    ResumptionAwaitingHandling frozen_res{
-        app.app_id(), ext, resumption_request};
-    frozen_resumptions_.push_back(frozen_res);
+  PendingRequest pending_request(app.app_id(), corr_id);
+  pending_requests_.push_back(pending_request);
+  SDL_LOG_DEBUG("Add to pending resumptins corr_id = " << corr_id);
+
+  if (pending_requests_.size() == 1) {
+    SendHMIRequestForNotSubscribed(pending_requests_.front());
   }
   resumption_data_processor().SubscribeToResponse(app.app_id(),
                                                   resumption_request);
 }
 
 void WayPointsPendingResumptionHandler::OnResumptionRevert() {
   SDL_LOG_AUTO_TRACE();
-  using namespace application_manager;
+  sync_primitives::AutoLock lock(pending_resumption_lock_);
 
-  if (!pending_requests_.empty()) {
-    SDL_LOG_DEBUG("Still waiting for some response");
+  if (pending_requests_.empty()) {
+    SDL_LOG_DEBUG("No pending resumptions");
     return;
   }
 
-  if (!frozen_resumptions_.empty()) {
-    ResumptionAwaitingHandling frozen_resumption = frozen_resumptions_.back();
-    frozen_resumptions_.pop_back();
-
-    auto request = std::make_shared<smart_objects::SmartObject>(
-        frozen_resumption.request_to_send_.message);
-    const uint32_t cid =
-        (*request)[strings::params][strings::correlation_id].asUInt();
-    const auto fid = static_cast<hmi_apis::FunctionID::eType>(
-        (*request)[strings::params][strings::function_id].asInt());
-
-    SDL_LOG_DEBUG("Subscribing for event with function id: "
-                  << fid << " correlation id: " << cid);
-    subscribe_on_event(fid, cid);
-    pending_requests_[cid] = *request;
-    SDL_LOG_DEBUG("Sending request with fid: " << fid << " and cid: " << cid);
-    application_manager_.GetRPCService().ManageHMICommand(request);
+  auto& pending_request = pending_requests_.front();
+  if (pending_request.waiting_for_hmi_response_) {
+    SDL_LOG_DEBUG("Pending resumption for  "
+                  << pending_request.app_id_
+                  << " is already waiting for HMI response");
+    return;
   }
+  SendHMIRequestForNotSubscribed(pending_request);
```

Comment:
I would recommend renaming to 
`SendNextPendingHMIRequest`, can we take the `pending_request` inside this function?

Reply:
I think that we don't need to do so, because inside of this function we use saved corr_id from initial request of processed app. If we don't take this request by reference as parameter, we will need to take it inside of this function with pending_requests_.front() method, but I think that it may cause some confusion with order of pending requests, especially taking in account possible recursive calls of ProcessNextPendingResumption method